core/src/main/scala/org/bykn/bosatsu/Evaluation.scala

```diff
@@ -199,6 +199,15 @@ object Evaluation {
     def unreachable: Scoped =
       const(Eval.later(sys.error("unreachable reached")))
 
+    def recursive(name: String, item: => Scoped): Scoped = {
```

Comment:
I don't think item actually has to be call-by-name. I think the laziness we need is the inner code recursively sets up the environment.